src/main/java/org/sagebionetworks/web/client/widget/ReactComponentV2.java

```diff
@@ -0,0 +1,277 @@
+import org.sagebionetworks.web.client.jsinterop.ReactComponentType;
+import org.sagebionetworks.web.client.jsinterop.ReactDOM;
+import org.sagebionetworks.web.client.jsinterop.ReactDOMRoot;
+import org.sagebionetworks.web.client.jsinterop.ReactElement;
+
+/**
+ * Abstract widget that manages the lifecycle of a {@link React} component tree mounted with {@link ReactDOM}.
+ * <p>
+ * To use it, extend the class, supply a {@link ReactComponentType} and {@link ReactComponentProps} to the constructor,
+ * and call {@link #render()} to render the React component.
+ * <p>
+ * This widget also manages appending child elements if the associated React component can contain children. If all
+ * child widgets implement this class, then the child {@link ReactElement}s will be passed as children to the React
+ * component. If any child of this component is a non-ReactComponent widget, then all child widgets (including
+ * implementations of {@link ReactComponentV2}) will be injected into the node found using the component's `ref`.
+ * <p>
+ * The root element defaults to a `div`, but can be changed (e.g. to a `span`) using the
+ * {@link #ReactComponentV2(T, P, String)} constructor.
+ * <p>
+ * This widget automatically unmounts the ReactComponent (if any) when this container is detached/unloaded.
+ */
+public abstract class ReactComponentV2<
+  T extends ReactComponentType<P>, P extends ReactComponentProps
+>
+  extends ComplexPanel
+  implements HasClickHandlers {
+
+  private ReactDOMRoot root;
+  private final ReactComponentType<P> reactComponentType;
+  public P props;
+
+  private final ArrayList<Element> nonReactChildElements = new ArrayList<>();
+
+  public ReactComponentV2(T reactComponentType, P props) {
+    this(reactComponentType, props, DivElement.TAG);
+  }
+
+  public ReactComponentV2(T reactComponentType, P props, String tag) {
+    this.reactComponentType = reactComponentType;
+    this.props = props;
+    setElement(Document.get().createElement(tag));
+  }
+
+  /**
+   * If any children of this widget do not implement this class, then we will inject them into the DOM using the React
+   * component's `ref` prop. This method will update the props with a callback ref that handles appending the children
+   * to the DOM node on which the ref is forwarded.
+   */
+  private void maybeUpdatePropsWithCallbackRef() {
+    if (!this.allChildrenAreReactComponents() && getChildren().size() > 0) {
+      // Create a callback ref that will allow us to inject the GWT children into the DOM
+      ReactComponentProps.CallbackRef callbackRef = (Element node) -> {
+        if (node != null) {
+          // Once the DOM node is defined, inject each child
+          getChildren()
+            .forEach(w -> {
+              node.appendChild(w.getElement());
+              // Keep track of child elements to ensure they are removed when the component unloads or re-renders
+              nonReactChildElements.add(w.getElement());
```

Comment:
Key change: after a non-React widget is injected into a React component, keep track of the element so we can properly detach it when this widget unloads.